dipy/reconst/shm.py

```diff
@@ -230,164 +232,205 @@ def real_sph_harm(m, n, theta, phi):
+    where the real harmonic $Y^m_n$ is defined to be:
 
-        Real($Y^m_n$)       if m > 0
-        $Y^0_n$             if m = 0
-        Imag($Y^|m|_n$)     if m < 0
+        Imag($Y^m_n$) * sqrt(2)      if m > 0
+        $Y^0_n$                      if m = 0
+        Real($Y^m_n$) * sqrt(2)      if m < 0
 
     This may take scalar or array arguments. The inputs will be broadcasted
     against each other.
 
     Parameters
-    -----------
+    ----------
     sh_order : int
         The maximum degree or the spherical harmonic basis.
     theta : float [0, pi]
         The polar (colatitudinal) coordinate.
     phi : float [0, 2*pi]
         The azimuthal (longitudinal) coordinate.
+    is_full_basis: bool, optional
+        If true, returns a basis including odd order SH functions as well as
+        even order SH functions. Otherwise returns only even order SH
+        functions. Default: False
+    is_legacy: bool, optional
+        If true, uses DIPY's legacy descoteaux07 implementation (where |m|
+        for m < 0). Else, implements the basis as defined in Descoteaux et al.
+        2007. Default: False
 
     Returns
-    --------
-    y_mn : real float
-        The real harmonic $Y^m_n$ sampled at `theta` and `phi` as
-        implemented in mrtrix. Warning: the basis is Tournier et al.
-        2007 [2]_; 2004 [1]_ is slightly different.
+    -------
+    real_sh : real float
+        The real harmonic $Y^m_n$ sampled at `theta` and `phi`.
     m : array
-        The order of the harmonics.
-    n : array
         The degree of the harmonics.
+    n : array
+        The order of the harmonics.
 
     References
     ----------
-    .. [1] Tournier J.D., Calamante F., Gadian D.G. and Connelly A.
-           Direct estimation of the fibre orientation density function from
-           diffusion-weighted MRI data using spherical deconvolution.
-           NeuroImage. 2004;23:1176-1185.
-    .. [2] Tournier J.D., Calamante F. and Connelly A. Robust determination
-           of the fibre orientation distribution in diffusion MRI:
-           Non-negativity constrained super-resolved spherical deconvolution.
-           NeuroImage. 2007;35(4):1459-1472.
-
+     .. [1] Descoteaux, M., Angelino, E., Fitzgibbons, S. and Deriche, R.
+           Regularized, Fast, and Robust Analytical Q-ball Imaging.
+           Magn. Reson. Med. 2007;58:497-510.
     """
-    m, n = sph_harm_full_ind_list(sh_order)
+    if is_full_basis:
+        m, n = sph_harm_full_ind_list(sh_order)
+    else:
+        m, n = sph_harm_ind_list(sh_order)
+
     phi = np.reshape(phi, [-1, 1])
     theta = np.reshape(theta, [-1, 1])
 
-    m = -m
-    real_sh = real_sph_harm(m, n, theta, phi)
-    real_sh /= np.where(m == 0, 1., np.sqrt(2))
-    return real_sh, m, n
+    if is_legacy:
+        # In the case where m < 0, legacy descoteaux basis
+        # consider |m| for the SH degree
+        deprecate_with_version('The legacy descoteaux07 SH basis is outdated \
```

Comment:
same comment as above.

> This is a decorator's initialization. The call is incorrect. Can you add a test to check this? You can look at `test_decorator.py` to see how to call the decorator.

Reply:
And again -- the same question: are there other functions in DIPY itself that need the legacy implementation of this function?